### PR TITLE
Fix #11254: "coqtop --version" working even in the middle of an installation process

### LINKED ISSUE #11254: ./bin/coqc --version no longer works on non-local builds
#### Description of the problem

This breaks my packaging scripts.

```
$ ./configure -prefix "$HOME/.local64/coq/coq-8.11+beta1/"
You have OCaml 4.06.1. Good!
You have OCamlfind 1.8.1. Good!
You have native-code compilation. Good!
You have the Num library installed. Good!
LablGtk3 not found:
=> no CoqIde will be built.

  Architecture                : Linux
  Sys.os_type                 : Unix
  Coq VM bytecode link flags  : -dllib -lcoqrun -dllpath /home/jgross/.local64/coq/coq-8.11+beta1//lib/coq/kernel/byterun
  Other bytecode link flags   :
  OCaml version               : 4.06.1
  OCaml binaries in           : /home/jgross/.opam/4.06.1+fp/bin/
  OCaml library in            : /home/jgross/.opam/4.06.1+fp/lib/ocaml
  OCaml flambda flags         :
  Native dynamic link support : true
  CoqIde                      : no
  Documentation               : None
  Web browser                 : firefox -remote "OpenURL(%s,new-tab)" || firefox %s &
  Coq web site                : http://coq.inria.fr/
  Bytecode VM enabled         : true
  Native Compiler enabled     : true

  Paths for true installation:
  - the Coq binaries will be copied in /home/jgross/.local64/coq/coq-8.11+beta1//bin
  - the Coq library will be copied in /home/jgross/.local64/coq/coq-8.11+beta1//lib/coq
  - the Coqide configuration files will be copied in /home/jgross/.local64/coq/coq-8.11+beta1//etc/xdg/coq
  - the Coqide data files will be copied in /home/jgross/.local64/coq/coq-8.11+beta1//share/coq
  - the Coq man pages will be copied in /home/jgross/.local64/coq/coq-8.11+beta1//share/man
  - the Coq documentation will be copied in /home/jgross/.local64/coq/coq-8.11+beta1//share/doc/coq
  - the Coqdoc LaTeX files will be copied in /home/jgross/.local64/coq/coq-8.11+beta1//share/texmf/tex/latex/misc

If anything is wrong above, please restart './configure'.

*Warning* To compile the system for a new architecture
          don't forget to do a 'make clean' before './configure'.
$ make states -j6
[...]
$ ./bin/coqc --version
Error: cannot guess a path for Coq libraries; please use -coqlib option

$ ./bin/coqc --print-version
Error: cannot guess a path for Coq libraries; please use -coqlib option


```

Consider 8.10.2 by contrast:
```
$ ./configure -prefix "$HOME/.local64/coq/coq-8.10.2/"
You have OCaml 4.06.1. Good!
You have OCamlfind 1.8.1. Good!
You have native-code compilation. Good!
You have the Num library installed. Good!
LablGtk3 not found:
=> no CoqIde will be built.

  Architecture                : Linux
  Sys.os_type                 : Unix
  Coq VM bytecode link flags  : -dllib -lcoqrun -dllpath /home/jgross/.local64/coq/coq-8.10.2//lib/coq/kernel/byterun
  Other bytecode link flags   :
  OCaml version               : 4.06.1
  OCaml binaries in           : /home/jgross/.opam/4.06.1+fp/bin/
  OCaml library in            : /home/jgross/.opam/4.06.1+fp/lib/ocaml
  OCaml flambda flags         :
  Native dynamic link support : true
  CoqIde                      : no
  Documentation               : None
  Web browser                 : firefox -remote "OpenURL(%s,new-tab)" || firefox %s &
  Coq web site                : http://coq.inria.fr/
  Bytecode VM enabled         : true
  Native Compiler enabled     : true

  Paths for true installation:
  - the Coq binaries will be copied in /home/jgross/.local64/coq/coq-8.10.2//bin
  - the Coq library will be copied in /home/jgross/.local64/coq/coq-8.10.2//lib/coq
  - the Coqide configuration files will be copied in /home/jgross/.local64/coq/coq-8.10.2//etc/xdg/coq
  - the Coqide data files will be copied in /home/jgross/.local64/coq/coq-8.10.2//share/coq
  - the Coq man pages will be copied in /home/jgross/.local64/coq/coq-8.10.2//share/man
  - the Coq documentation will be copied in /home/jgross/.local64/coq/coq-8.10.2//share/doc/coq
  - the Coqdoc LaTeX files will be copied in /home/jgross/.local64/coq/coq-8.10.2//share/texmf/tex/latex/misc

If anything is wrong above, please restart './configure'.

*Warning* To compile the system for a new architecture
          don't forget to do a 'make clean' before './configure'.
$ make states -j6
[...]
$ ./bin/coqc --version
The Coq Proof Assistant, version 8.10.2 (December 2019)
compiled on Dec 6 2019 23:12:12 with OCaml 4.06.1
$ ./bin/coqc --print-version
8.10.2 4.06.1
```

#### Coq Version

8.11+beta1

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix

Fixes #11254  

As a side remark, the semantics of the command line remains a bit unsatisfactory in general. For instance, is it obvious that we'd like `bin/coqtop -coqlib foo -where -coqlib bar` to return `bar`. I wonder if the list of options shouldn't be either clearly processed sequentially (like in a v-file) or, at worst, clearly preventing redundancy (see also #9585).

- [X] Entry added in the changelog

### CODE REVIEW

File: doc/changelog/08-tools/11255-master+fix11254-coqtop-version.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+- **Fixed:**
+  `coqtop --version` working even in the middle of an installation process
```
Comment (Zimmi48) [inline doc/changelog/08-tools/11255-master+fix11254-coqtop-version.rst]: Since there is the word "fixed" above, I'd rather formulate this negatively:

```suggestion
  `coqtop --version` not working in the middle of an installation process
```

Comment (herbelin) [reply doc/changelog/08-tools/11255-master+fix11254-coqtop-version.rst]: "`coqtop --version` was broken when called in the middle of an installation process" ??

Comment (Zimmi48) [reply doc/changelog/08-tools/11255-master+fix11254-coqtop-version.rst]: Even better.

File: doc/changelog/08-tools/11255-master+fix11254-coqtop-version.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+- **Fixed:**
+  `coqtop --version` was broken when called in the middle of an installation process
+  (`#11255 <https://github.com/coq/coq/pull/11255>`_, by Hugo Herbelin, fixing
+   `#11254 <https://github.com/coq/coq/pull/11254>`_).
```
Comment (ejgallego) [inline doc/changelog/08-tools/11255-master+fix11254-coqtop-version.rst]: It seems indentation here broke the refman build

File: lib/envars.ml [hunk 3]
```diff
@@ -171,11 +171,15 @@ let xdg_dirs ~warn =
...
+  let n = String.length s in
+  if n <> 0 && s.[n-1] <> '/' then s ^ "/" else s
+
 let print_config ?(prefix_var_name="") f coq_src_subdirs =
   let open Printf in
   fprintf f "%sLOCAL=%s\n" prefix_var_name (if Coq_config.local then "1" else "0");
-  fprintf f "%sCOQLIB=%s/\n" prefix_var_name (coqlib ());
-  fprintf f "%sDOCDIR=%s/\n" prefix_var_name (docdir ());
+  fprintf f "%sCOQLIB=%s\n" prefix_var_name (ensure_trailing_slash (coqlib ()));
+  fprintf f "%sDOCDIR=%s\n" prefix_var_name (ensure_trailing_slash (docdir ()));
```
Comment (ejgallego) [inline lib/envars.ml]: This change seems unrelated, right? Also, is it correct? IIUC these variables are not supposed to have a trailing slash.

**Review by ejgallego — COMMENTED** [review]
Not sure the original bug report is valid IMHO, if you are calling the binaries in an inconsistent state not sure it should work.

Even if we choose to fix that this way I guess we should not duplicate the call to `init_setup`, but indeed refine on the needed state?

**Comment by SkySkimmer** [comment]
>Not sure the original bug report is valid IMHO, if you are calling the binaries in an inconsistent state not sure it should work.

I think --version at least is special enough that it should always work (to help bug reports for instance)

File: lib/envars.ml (hunk 3 above)
Comment (herbelin) [reply lib/envars.ml]: This is unrelated to the title of the PR. It is related to what `coqc -config` reports (formerly, I observed things like `/mydir//`). It is maybe too minor for being worth a different PR.

File: doc/changelog/08-tools/11255-master+fix11254-coqtop-version.rst (hunk 2 above)
Comment (herbelin) [reply doc/changelog/08-tools/11255-master+fix11254-coqtop-version.rst]: I mixed up the syntax. It is supposed to be ``` ``coqtop --version`` ```. Fixed.

**Comment by herbelin** [comment]
@skyskimmer, @ejgallego:

There is indeed a pending question of what semantics to give to the command line in general.

For instance, what are the following commands supposed to do: `coqc ackermann_4_4.v --version`, or `coqc --version ackermann_4_4.v`, or `coqc -where --version`, or `coqc --version -where`?

Should we forbid `--version` in combination with other arguments? Should we process options sequentially? Should we process options in several phases, for instance interpreting `--version` in the very first phase? Etc.

File: lib/envars.ml (hunk 3 above)
Comment (ejgallego) [reply lib/envars.ml]: Umm, this doesn't affect what `coqc -config` outputs, but only what gets written for coq_makefile, right?

The code also doesn't seem correct, I dunno what problem is trying to solve but that should be done then at the level of `coqlib ()`, not here in the printing. I'd suggest we remove this change for now, as this PR has to be backported.

File: doc/changelog/08-tools/11255-master+fix11254-coqtop-version.rst (hunk 2 above)
Comment (herbelin) [reply doc/changelog/08-tools/11255-master+fix11254-coqtop-version.rst]: Indentation was wrong also. Also fixed.

File: lib/envars.ml (hunk 3 above)
Comment (herbelin) [reply lib/envars.ml]: > this doesn't affect what coqc -config outputs, but only what gets written for coq_makefile, right?

It is called in `toplevel/coqtop.ml`. I believe it was used in former versions of `coq_makefile`.

I don't know what are the exact specifications of `coqlib ()` and `docdir ()`, i.e. I don't know if they are intended to return a path ended by `/` or not.

> I'd suggest we remove this change for now, as this PR has to be backported.

OK. I'm removing it.

Comment (ejgallego) [reply lib/envars.ml]: > I don't know what are the exact specifications of `coqlib ()` and `docdir ()`, i.e. I don't know if they are intended to return a path ended by `/` or not.

This changed recently; indeed it is always confusing. Sometime ago I tried to follow the approach used in Dune of having an opaque `Path.t` type, but indeed that's pretty challenging to do, on the other hand it brings consistency.

**Review by ejgallego — APPROVED** [review]
Thanks @herbelin for the fix and sorry for the delay over nits; the solution is OK for now, I'll merge as soon as CI has passed as this needs to be backported.
